### PR TITLE
Add support for passing severity to `report`

### PR DESCRIPTION
**Pull request by kangax**

### CODE REVIEW

File: src/utils/report.js [hunk 1]
```diff
@@ -52,11 +53,11 @@ export default function ({
     }
   }
 
-  const severity = (result.stylelint.ruleSeverities)
+  const actualSeverity = (result.stylelint.ruleSeverities)
```
Comment (kangax) [inline src/utils/report.js]: Not sure about this name; suggestions are welcome :)

File: src/utils/report.js [hunk 2]
```diff
@@ -52,11 +53,11 @@ export default function ({
     }
   }
 
-  const severity = (result.stylelint.ruleSeverities)
+  const actualSeverity = severity || (result.stylelint.ruleSeverities
```
Comment (kangax) [inline src/utils/report.js]: Not sure about this name; suggestions welcome :

**Comment by ngavalas** [comment]
So, severity can already be changed from error to warning by setting the warn option when you're specifying the rules:
https://github.com/stylelint/stylelint/blob/3f2fbc1d31863cf8744454c3a2f38505de2b0928/docs/user-guide/configuration.md#severities-error--warning

For some context, the real reason that @kangax wants this is because we want 'advice' and 'warning' instead of 'warning' and 'error'.  Perhaps the better option is to add a `severity` flag to each rule (instead of `warn`, as we have now) instead of making report() care about this.

**Comment by kangax** [comment]
@ngavalas Good point. I feel like having "advice" type in stylelint could make sense for general audience as well.

File: src/utils/report.js [hunk 3]
```diff
@@ -52,11 +53,11 @@ export default function ({
...
     ? result.stylelint.ruleSeverities[ruleName]
```
Comment (kangax) [inline src/utils/report.js]: I actually did try setting this manually but it didn't work (presumably because it gets set implicitly via rule option)

**Comment by davidtheclark** [comment]
Is the purpose of the "advice" severity just to customize the appearance in the formatter?

I wouldn't be opposed to adding an "advice"-level severity. It makes me glad we dropped the 0-1-2 system; because a new severity level would work into our API simply by adding a secondary option (instead of `indentation: [2, { warn: true }]` could have `indentation: [2, { advice: true }]`).

Besides that, I don't really understand the intention of the PR. Why do you want the rule to pass severity?

**Comment by kangax** [comment]
> Why do you want the rule to pass severity?

Well, rule defines logic, message, explanation, and severity. Then I pass all of that to report. I'll try to figure our custom formatters now as it seems to be what we're looking for here.

**Comment by davidtheclark** [comment]
@kangax : "Well, rule defines logic, message, explanation, and severity." --- I still feel like I don't get it. The *user* defines the severity, right? Not the rule? What is it that you want stylelint to do that it currently doesn't do -- could you describe a use-case? But maybe this is a moot point if you're finding another way?

**Comment by kangax** [comment]
@davidtheclark My use case is that we have custom rules, defined via plugins. Some of the rules that we use are already in Stylelint (like !important declaration), but others are either specific to our app or specific to certain browsers or are mere suggestions or are simply not in a Stylelint yet. For each custom rule I have a file now — a plugin. There, I define what that rule should do and what it should report when it "hits" the condition. For example, we might want to issue warning for incompatible "flex" values or issue an "advice" (aka caution) when certain perf-senstive properties are used (opacity, transform, etc.). When I'm calling `report` from within a plugin (aka rule), I'm passing text, lineNumber, etc. and I would also like to pass "severity" to be reported; or really just to be passed along. I thought it makes sense since severity is already passed to the output from within `report`. It just can't be "set" externally via report. 

Hope it makes more sense now.

**Comment by davidtheclark** [comment]
@kangax Sure, thanks. The pattern that's going right now is that the rules themselves don't know their severities. They just call `report()` to complain, and `report()` knows about severities that were registered in the configuration. All rules, even plugin rules, have their severities defined in the configuration. So I think that the rule shouldn't have to (or be able to) pass severity to `report()`: severities are always specified in the config; the rules themselves don't know or care about their severities; and `report()` deals with severity when it is called.

Does that make sense? Or are you having problems setting the severity for your plugins via the config object?

**Comment by kangax** [comment]
Fair enough, I'll try it like that then. Thanks!

**Pull request closed without merging**